### ISSUE TITLE
Enhancement: Scroll to top of page (i.e., top of <body> element)

### ISSUE THREAD

**Issue by alex-lew** [elm-lang/dom#9]
Currently, if you wish to issue a command to scroll to the top of the page, I believe you need to 

- create a custom `index.html` where the `<body>` tag has an `id` attribute
- compile to JS instead of HTML (giving up tools like `elm-reactor`)
- issue a `Dom.Scroll.toTop "bodyId"` command

In my work, wanting to scroll to the top of a page has been relatively common. That's because Elm apps, even when they conceptually contain "multiple pages," are single-page from the browser's perspective. Scrolling down to the footer of some Elm app and clicking "About us" may cause the view to change, but will not cause the browser to reset the scroll position. So you end up at the bottom of the new About us page, instead of the top.

I'm not sure if this is a common enough use case, or if there's a better solution than using `Dom.Scroll` here, but when I asked on Slack, someone suggested I post an issue here. It's come up enough for me that I wonder if it might be nice to have a special command, exposed by `Dom.Scroll`, that scrolls to the top of the `body` element. That way, this functionality could be available without adding a custom `index.html` and foregoing `elm-reactor`.

Thanks!

**Comment by cbenz** [comment]
You don't really have to give up `elm-reactor`: just create your custom `index.html` based on the one `elm-reactor` displays by default, then open `http://localhost:8000/index.html`.

That's what I do in a personal project [noteminer](https://github.com/cbenz/noteminer), see its [`index.html`](https://github.com/cbenz/noteminer/blob/master/index.html).

But this issue is very relevant.

I think I have a solution:
- adding a `type DocumentElement = Html | Body | ById String`
- modifying the `Dom` functions to accept a `DocumentElement` type instead of an id only (breaking changes everywhere)
- implementing the change in the [`withNode` function](https://github.com/elm-lang/dom/blob/22e3f2f4170e441d318e85a380379655ffeb53c6/src/Native/Dom.js#L48) to use:
    - `document.documentElement` for the `Html` constructor
    - `document.body` for the `Body` constructor
    - `getElementById` for the `ById` constructor

Shall I try to do a pull request in that direction? Or is there a way to avoid introducing a new type and break the `Dom` API?

I'd like to know what the authors @evancz and @rtfeldman think about it before :-)

**Comment by rriegs** [comment]
I'd add that `Dom.Scroll` should offer a specific method like `toTopOfPage` that issues both `toTop Html` and `toTop Body` for maximum portability: Safari apparently uses the `body` element to manage page scroll while the other major browsers all use `documentElement`.

**Comment by rtfeldman** [comment]
> Safari apparently uses the `body` element to manage page scroll while the other major browsers all use `documentElement`.

In that case, wouldn't it be better for a hypothetical `Dom.scrollToPageTop` function to automatically do the right thing for the current browser?

**Comment by rriegs** [comment]
I agree, and that is my motivation for `toTopOfPage`: it's too easy for users to forget that just one of `toTop Body` and `toTop Html` is insufficient for all browsers (or possibly even all future versions of their own browser).

The code you link (and [other code examples I can find](https://www.w3schools.com/howto/howto_js_scroll_to_top.asp)) behave equivalently to issuing both of the proposed `toTop Body` and `toTop Html` in sequence, which is what I call for as well.

I don't think it's worth attempting to detect the current browser software and version in order to determine which to perform when it is always safe to issue both.  This practice should continue to work even as new browsers/versions are released that might require one or the other method.  If, on the other hand, some future browser requires an entirely different method to scroll to top, the motivation for a separate `toTopOfPage` is even stronger: existing Elm websites could then be made to support this aspect of the new browser by simply upgrading their elm-lang/dom dependency.

**Comment by muelli** [comment]
> create a custom index.html where the <body> tag has an id attribute
> compile to JS instead of HTML (giving up tools like elm-reactor)
> issue a Dom.Scroll.toTop "bodyId" command

That didn't work for me. `Dom.Scroll.toTop` sets `element.scrollTop` to 0 and somehow my `body`'s `scrollTop` is 0 even though I've scrolled all the way down. `scrollIntoView` helped me to scroll to the element of my desire. There is a patch rotting here: https://github.com/elm-lang/dom/pull/13/files

So far I've helped myself using a port:

```
port module ScrollIntoView exposing (..)

{-|
  Call scrollIntoView() on the element with the given Id.

  See https://developer.mozilla.org/de/docs/Web/API/Element/scrollIntoView
-}

port scrollIntoView : String -> Cmd msg
```


With this TypeScript code:
```
    app.ports.scrollIntoView.subscribe( (elem) => {
        // requestAnimationFrame( () => {
            const element = document.getElementById(elem);
            if (element == null) {
               throw new Error("Element not found: " + elem);
            } else {
               // if (element.scrollIntoViewIfNeeded != null) {
               //   element.scrollIntoViewIfNeeded(false);
               // } else {
               element.scrollIntoView();
               // }
           }
   });
```
I'm confused as to when the port is being called, i.e. after the view has been rendered or just before, so I've left the option of delaying the scroll after the next frame in place. But it worked for me without that.
